File: LeetCode/Python/src/Substring_With_Concatenation_Of_All_Words.py
"""
@author - Anirudh Sharma
"""
from typing import List


class SubstringWithConcatenationOfAllWords:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        # Resultant list
        indices = []
        # Base conditions
        if s is None or len(s) == 0 or words is None or len(words) == 0:
            return indices

        # Dictionary to store the count of each word in the words array
        wordCount = dict()
        # Loop to store count of each word in the array
        for i in range(len(words)):
            if words[i] in wordCount:
                wordCount[words[i]] += 1
            else:
                wordCount[words[i]] = 1
        # Length of each word in the words array
        wordLength = len(words[0])
        # Total length of all the words in the array
        wordArrayLength = wordLength * len(words)
        # Loop for each character in the string
        for i in range(0, len(s) - wordArrayLength + 1):
            # Get the current string
            current = s[i:i + wordArrayLength]
            # Map to store the count of each word in the current
            wordMap = dict()
            # Index to loop through the array
            index = 0
            # Index to partition the current string
            j = 0
            # Loop through the words array
            while index < len(words):
                # Partition the string
                part = current[j: j + wordLength]
                # Save this in wordMap
                if part in wordMap:
                    wordMap[part] += 1
                else:
                    wordMap[part] = 1
                # Update the indices
                j += wordLength
                index += 1
            # Compare the two maps
            if wordMap == wordCount:
                indices.append(i)
        return indices


if __name__ == '__main__':
    sObject = SubstringWithConcatenationOfAllWords()
    s = "barfoothefoobarman"
    words = ["foo", "bar"]
    print(sObject.findSubstring(s, words))

    s = "wordgoodgoodgoodbestword"
    words = ["word", "good", "best", "word"]
    print(sObject.findSubstring(s, words))

    s = "barfoofoobarthefoobarman"
    words = ["bar", "foo", "the"]
    print(sObject.findSubstring(s, words))

    s = "wordgoodgoodgoodbestword"
    words = ["word", "good", "best", "good"]
    print(sObject.findSubstring(s, words))

File: LeetCode/JavaScript/src/substring_with_concatenation_of_all_words.js
var findSubstring = function (s, words) {
    // Resultant list
    const indices = [];
    // Base conditions
    if (s === null || s.length === 0 || words === null || words.length == 0) {
        return indices;
    }
    // Store the words and their counts in a hash map
    const wordCount = words.reduce((a, b) => {
        a[b] = (a[b] + 1) || 1;
        return a;
    }, {});
    // Length of each word in the words array`
    const wordLength = words[0].length;
    // Length of all the words combined in the array
    const wordArrayLength = wordLength * words.length;
    // Loop for the entire string
    for (let i = 0; i <= s.length - wordArrayLength; i++) {
        // Get the substring of length equal to wordArrayLength
        let current = s.substring(i, i + wordArrayLength);
        // Map to store each word of the substring
        const wordMap = {};
        // Index to loop through the words array
        let index = 0;
        // Index to get each word in the current
        let j = 0;
        // Loop through each word of the words array
        while (index < words.length) {
            // Divide the current string into strings of length of
            // each word in the array
            const part = current.substring(j, j + wordLength);
            // Put this string into the wordMap
            wordMap[part] = (wordMap[part] + 1) || 1;
            // Update j and index
            j += wordLength;
            index++;
        }
        // At this point compare the maps
        let a = JSON.stringify(Object.entries(wordCount).sort());
        let b = JSON.stringify(Object.entries(wordMap).sort());
        if (a === b) {
            indices.push(i);
        }
    }
    return indices;  
};

let str = "barfoothefoobarman";
let wordsArray = ["foo", "bar"];
console.log(findSubstring(str, wordsArray));

str = "wordgoodgoodgoodbestword"
wordsArray = ["word", "good", "best", "word"]
console.log(findSubstring(str, wordsArray));

str = "barfoofoobarthefoobarman"
wordsArray = ["bar", "foo", "the"]
console.log(findSubstring(str, wordsArray));

str = "wordgoodgoodgoodbestword"
wordsArray = ["word", "good", "best", "good"]
console.log(findSubstring(str, wordsArray));